### ISSUE TITLE
Add a continuous saturation genome editing ranking task

### ISSUE THREAD

**Issue by gonzalobenegas** [Open-Athena/VEP-bench#48]
# Add a continuous saturation genome editing ranking task

## Summary

Add an `sge` benchmark task built directly from pinned, public saturation genome
editing score sets rather than reusing the derived `marin-dna/evals_sge`
artifact. Each question should contain one gene, one deterministically selected
exon with 100 bp of unmarked flanking sequence on each side, and 50 assayed
SNVs. The model predicts one continuous functional-damage score per SNV.

Reuse VEPBench's existing `ranking` question type and the output, parsing, and
scoring conventions already established for
[satMutMPRA](docs/tasks/satmut-mpra.md). Mean within-gene Spearman correlation
is the primary metric; mean within-gene Pearson correlation and valid-output
rate are secondary metrics.

## Motivation

SGE measures variant effects at endogenous genomic loci and therefore
complements the reporter-based regulatory effects in satMutMPRA. A useful task
should preserve the continuous experimental signal, include both protein-coding
and local splice-disrupting variation, and expose the molecular and assay
context needed to define the prediction target without leaking consequence
annotations or source identifiers.

Rebuilding from the primary score sets also lets us:

- audit the current MaveDB catalog, including deposits newer than the existing
  derived artifact;
- choose one canonical experiment and one question per gene;
- pin source payloads and all curation decisions;
- apply a task-specific exon and 50-variant sampling policy; and
- keep the complete eligible population in a reusable preparation cache.

## Task contract

- **Task family:** `sge`
- **Question type:** existing `ranking`
- **Question unit:** one selected exon from one gene
- **Questions per gene:** exactly one
- **Candidates per question:** exactly 50 SNVs
- **Prediction target:** continuous functional damage, with larger values meaning
  greater damage
- **Primary metric:** arithmetic mean of within-gene Spearman correlations, with
  every included gene weighted equally
- **Secondary metrics:** arithmetic mean of within-gene Pearson correlations and
  valid-output rate

Do not pool raw scores across genes. Assay score scales differ, and both
correlations are computed within each question before macro-averaging.

## Current MaveDB catalog

The initial catalog audit was performed on 2026-09-03 through the public
[MaveDB API](https://www.mavedb.org/docs/mavedb/programmatic-access/api-quickstart.html).
The following are the provisional canonical score sets. The preparation step
must still verify that each gene has an exon with at least 50 eligible variants;
genes that fail the complete coordinate, consequence, QC, or capacity checks
are excluded and recorded in the manifest.

| Gene | Provisional score set | Notes |
| --- | --- | --- |
| BAP1 | [`urn:mavedb:00000662-0-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00000662-0-1) | Aggregate SGE score set |
| BARD1 | [`urn:mavedb:00001250-a-2`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001250-a-2) | Current aggregate score set |
| BRCA1 | [`urn:mavedb:00000097-0-2`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00000097-0-2) | Normalized aggregate across exon experiments |
| BRCA2 | [`urn:mavedb:00001279-a-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001279-a-1) | HAP1 SGE of exons 15-26; retained deliberately |
| CTCF | [`urn:mavedb:00001262-a-3`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001262-a-3) | Current aggregate score set |
| DDX3X | [`urn:mavedb:00001228-a-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001228-a-1) | Direct experimental screen; prefer over rounded clinical-use scores |
| PALB2 | [`urn:mavedb:00001259-a-2`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001259-a-2) | Current aggregate score set |
| RAD51C | [`urn:mavedb:00000673-0-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00000673-0-1) | Aggregate SGE score set |
| RAD51D | [`urn:mavedb:00001260-a-2`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001260-a-2) | Current aggregate score set |
| SBDS | [`urn:mavedb:00001284-a-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001284-a-1) | New SGE deposit |
| SFPQ | [`urn:mavedb:00001265-a-2`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001265-a-2) | Current aggregate score set |
| TINF2 | [`urn:mavedb:00001286-a-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001286-a-1) | New SGE deposit covering exons 4-6 |
| TP53 | [`urn:mavedb:00001277-a-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001277-a-1) | Endogenous p53 DNA-binding-domain scan |
| VHL | [`urn:mavedb:00000675-a-1`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00000675-a-1) | Complete VHL SGE data |
| XRCC2 | [`urn:mavedb:00001264-a-2`](https://api.mavedb.org/api/v1/score-sets/urn:mavedb:00001264-a-2) | Current aggregate score set |

Select exactly one canonical score set per gene. Prefer, in order:

1. the latest non-superseded aggregate score set;
2. an endogenous-locus SGE experiment with a direct loss-of-function readout;
3. combined measurements over individual replicates or exon-specific deposits;
4. the baseline or standard functional condition over a drug-specific or
   gain-of-function condition; and
5. the primary experimental score over a rounded or clinically transformed
   derivative.

Explicit exclusions:

- Exclude CARD11. Although its loss-of-function arm is endogenous diploid SGE,
  the library is predominantly multi-base codon substitutions and provides
  little or no splice-variant representation after restricting to SNVs. Also
  exclude its gain-of-function and ibrutinib arms.
- Exclude duplicate per-exon, per-replicate, treatment-specific, and superseded
  score sets once a canonical aggregate has been selected.
- Exclude cDNA complementation, minigene, reporter, extra-local construct,
  guide-level, and unresolved-bystander assays. Editing chemistry is otherwise
  unrestricted when each score is attributable to one specific endogenous SNV.

The final implementation should rerun and record the catalog search, rather
than treating this provisional table as a substitute for a pinned source
manifest.

## Source preparation and provenance

Download each selected score set directly from MaveDB. Use its primary `score`
column as the reference value. Do not recompute scores from raw sequencing
counts and do not substitute author classes or pathogenicity calibrations for
the continuous experimental target.

For every MaveDB payload, pin and record:

- score-set URN, experiment URN, title, modification and retrieval dates;
- license, target gene, target accession or sequence, and publication links;
- payload URL, byte size, and SHA-256 digest;
- assay controlled vocabulary and the exact source column inventory; and
- the manually reviewed score direction and model-visible assay description.

Use a pinned GRCh38 primary assembly for sequence extraction and REF validation.
Use the assay's declared target transcript to define exon geometry and 5-prime
to 3-prime display orientation. If it is unavailable, use a pinned MANE Select
transcript as a deterministic fallback. The selected transcript defines only
the displayed window and orientation; it does not constrain the VEP consequence
annotation.

For transcript-level `c.` HGVS score sets, recover GRCh38 genomic coordinates
with the pinned `pyhgvs` plus `cdot` procedure used by the upstream SGE pipeline,
including intronic SNVs that are absent from MaveDB's mapped-variant output.
Genomic `g.` HGVS may be parsed directly. Reject unmapped or ambiguous records,
validate every REF against GRCh38, and do not repair an SGE record by swapping
REF and ALT. A swap changes which allele the experimental score describes and
must be treated as a source/build error.

Keep source and genomic coordinates, transcript, exon boundaries, consequence,
optional amino-acid annotation, QC fields, score direction, author metadata,
and selection metadata as private provenance. Amino-acid metadata is optional
and may be omitted when it is difficult to obtain reliably.

## SNV and consequence eligibility

Restrict candidates to unique substitutions where:

- REF and ALT are different single bases in `A/C/G/T`;
- the variant maps uniquely to GRCh38 and REF matches the pinned assembly;
- the primary experimental score is finite; and
- the source does not explicitly mark the variant as failed QC, low confidence,
  or indeterminate.

Use the pinned
[`songlab/hg38-variant-consequences`](https://huggingface.co/datasets/songlab/hg38-variant-consequences)
artifact at revision `eb3022cc6797b9369cca16af72ff3c4197df343a`.
It was generated with Ensembl VEP 109.1 using `--most_severe --distance 1000`.
The global most-severe consequence is authoritative even when it comes from a
transcript other than the transcript used to define the displayed exon. Do not
require transcript concordance.

Reimplement and provenance-pin the compact eligibility policy from the upstream
[SGE pipeline](https://github.com/Open-Athena/marin-dna/blob/0ce78a4effba482fd5b137058f9214cc5d8d6682/src/marin_dna/pipelines/evals/sge.py)
and
[configuration](https://github.com/Open-Athena/marin-dna/blob/0ce78a4effba482fd5b137058f9214cc5d8d6682/snakemake/evals/config/config.yaml):

1. Exclude the following most-severe consequences before recategorization:
   `transcript_ablation`, `splice_acceptor_variant`,
   `splice_donor_variant`, `stop_gained`, `frameshift_variant`, `stop_lost`,
   `start_lost`, `transcript_amplification`, `feature_elongation`, and
   `feature_truncation`.
2. Reclassify an `intron_variant` whose distance to the nearest annotated exon
   is at most 30 bp as `exon_proximal`.
3. Define the private `splicing` group as
   `splice_donor_5th_base_variant`, `splice_region_variant`,
   `splice_donor_region_variant`, `splice_polypyrimidine_tract_variant`, and
   `exon_proximal`.
4. Retain only `missense_variant` and `splicing` candidates.

Consequence class and distance are selection metadata and must not be shown to
the model.

## Score orientation

Orient every gene so a larger reference value means greater functional damage.
Apply only a sign flip when the source's primary score runs in the opposite
direction; do not otherwise normalize, standardize, calibrate, or rescale the
score. Document the direction decision using the assay methods, controls, and
source metadata. Exclude a gene if score direction cannot be established
unambiguously.

## Exon and panel selection

For each eligible gene:

1. Construct the eligible missense and splicing pools for each assayed exon and
   its 100 bp genomic flanks on both sides.
2. Discard exon windows with fewer than 50 eligible variants in total.
3. Prefer exons that can supply 25 missense and 25 splicing variants.
4. If none can, choose the exon that maximizes the achievable size of the
   smaller class, then fill the remaining positions from the other class. Exact
   25/25 balance is a preference, not an eligibility requirement.
5. Break remaining ties by the largest robust reference-score spread
   (`P95 - P05`), then by a stable transcript/exon coordinate key.
6. Within each class allocation, sample evenly across reference-score quantiles
   using a pinned SHA-256 ordering and seed, adapting the deterministic
   satMutMPRA rank-quantile procedure.
7. Sort the selected variants by displayed local position, REF, and ALT, then
   assign opaque IDs `V01` through `V50`.

Record the full eligible population counts, class counts, exon-selection
features, quantile bins, seed, tie-breaks, and selected records in the cache and
manifest. The public question must not reveal score quantiles or class balance.

## Model-visible prompt

Each prompt should expose:

- the gene symbol;
- a concise, manually reviewed description of the causally relevant assay
  context: editing design, cellular/model system, engineered background,
  selection or measurement, treatment when integral to the assay, timing or
  contrast, and how loss of function changes the measured phenotype;
- one continuous uppercase DNA sequence containing the selected exon plus
  exactly 100 bp on each side, displayed in transcript 5-prime to 3-prime
  orientation; and
- a compact VCF containing synthetic contig `element`, 1-based positions within
  the displayed sequence, opaque IDs, REF, and ALT.

For a negative-strand transcript, reverse-complement the complete window and
both alleles before computing the local positions. Do not mark exon boundaries
through capitalization, separators, coordinates, prose, or other formatting.

Do not expose:

- genomic coordinates, transcript identifiers, exon numbers, or exon
  boundaries;
- per-variant consequence class or amino-acid substitution;
- reference scores, QC fields, uncertainty, author classes, or sampling bins;
- publication titles, authors, PMIDs, MaveDB accessions, source-specific column
  names, or disease labels.

Do not describe every assay generically as "fitness." State its actual
experimental readout and the causal interpretation of depletion, enrichment,
proliferation, morphology, or treatment response as appropriate.

Ask for one finite numeric predicted-damage score for every candidate. Larger
predictions must mean greater damage; no cross-gene absolute calibration is
required. Reuse satMutMPRA's exact final-answer form:

```text
FINAL: {"V01": -0.42, "V02": 0.08, ...}
```

## Parsing and scoring

Reuse the existing ranking parser and scorer without task-specific heuristics:

- parse only the last well-formed `FINAL: {JSON object}` line;
- require every expected candidate ID exactly once and reject extra IDs;
- reject booleans, strings, nulls, nonnumeric values, and non-finite numbers;
- assign both correlations a value of zero for an invalid completed output;
- define a correlation as zero when either relevant vector is constant;
- keep API failures null and mark the run incomplete; and
- report valid-output rate alongside mean Spearman and mean Pearson.

The existing ranking framework already implements Spearman, Pearson, strict
JSON validation, invalid-output zeros, and macro aggregation. No new question
or result schema should be introduced unless implementation uncovers a genuine
incompatibility.

## Implementation outline

- Add a task-local package under `tasks/sge/`, modeled on
  `tasks/satmut-mpra/`, for catalog retrieval, source validation, coordinate
  recovery, consequence joins, curation, deterministic sampling, and artifact
  validation.
- Add strictly validated human-maintained preparation configuration containing
  one entry per provisional gene: canonical URN, transcript policy, expected
  chromosome, score direction, QC mapping, and model-visible assay context.
- Add `configs/tasks/sge/task.yaml` and a versioned SGE prompt template.
- Generate a compact source JSONL, source manifest, and expected question-set
  manifest. Do not hand-edit generated artifacts.
- Cache the complete processed eligible population and provenance separately
  from the compact 50-variant panels. Use content-addressed, immutable cache
  keys and publish the completion manifest last, following satMutMPRA.
- Join the very large consequence artifact per chromosome with predicate
  pushdown or an equivalently bounded method. Do not eagerly load or globally
  group/sort the complete all-SNV dataset on the local VM.
- Add task methodology under `docs/tasks/sge.md`, update the task catalog and
  relevant development/evaluation/publication guides, and add the task to the
  static results explorer.
- Add a remote preparation entry point when the full source build exceeds the
  shared-node limits. Evaluation remains an explicit local action and tests
  must not make live MaveDB, Hugging Face, or model API calls.

## Acceptance criteria

- [ ] A pinned catalog audit selects at most one score set per gene and records
      all included and excluded records with reasons.
- [ ] BRCA2 is included if it passes the same eligibility checks as every other
      gene; CARD11 is excluded by policy.
- [ ] Every included assay is endogenous-locus, saturation-scale, and has a
      score attributable to a single SNV.
- [ ] Every included gene produces exactly one deterministic 50-variant panel
      from one exon window; genes without a qualifying window are excluded and
      documented.
- [ ] Each displayed sequence contains exactly the selected exon plus 100 bp on
      each side, is entirely uppercase, is in transcript orientation, and does
      not mark exon boundaries.
- [ ] Candidate coordinates are 1-based and local to synthetic contig
      `element`; every visible REF agrees with the displayed sequence.
- [ ] Only eligible missense and defined splicing SNVs are sampled; canonical
      splice, nonsense, frameshift, and other configured high-impact variants
      are absent.
- [ ] The sampler prefers 25 missense and 25 splicing variants, relaxes this
      without excluding an otherwise valid gene, covers score quantiles within
      each class, and is byte-reproducible from pinned inputs and configuration.
- [ ] Ground-truth scores use the MaveDB primary `score` with direction-only
      alignment so larger values mean greater damage.
- [ ] Prompts expose the gene and relevant assay mechanism but no consequence,
      amino-acid, source-identity, genomic-coordinate, score, QC, or selection
      leakage.
- [ ] The generated questions validate against the existing ranking schema and
      the evaluator reports equal-weight mean Spearman, mean Pearson, and
      valid-output rate using the established satMutMPRA behavior.
- [ ] Source records, manifests, cache contents, and question generation are
      deterministic and digest-verified.
- [ ] Offline tests cover source/config validation, SNV filtering, coordinate
      orientation, REF agreement, the 30/31 bp splice boundary, consequence
      exclusion, score direction, exon selection, class allocation, quantile
      sampling, prompt leakage, manifests, and regeneration.
- [ ] Task, development, evaluation, publication, and static-site documentation
      are updated in the same change.

## Known limitations

- A single exon plus 100 bp of local context omits the rest of the gene,
  long-range regulation, distant splice context, and protein-domain context.
  The visible gene symbol partially mitigates this but does not replace the
  missing sequence.
- Global VEP `--most_severe` annotations can be driven by a transcript other
  than the assay transcript. This is an intentional, documented task definition
  rather than a concordance guarantee.
- Quantile sampling and consequence balancing create an informative evaluation
  panel but do not reproduce the natural distribution of experimental effects
  or variant classes.
- SGE measurements remain specific to their cell system, engineered background,
  selection, time point, and treatment. They measure experimental function, not
  clinical pathogenicity or universal organismal fitness.
- The questions and answers are a public development set and may occur in model
  training data.

## Out of scope

- binary labels or pathogenic/benign classification;
- nonsense, canonical splice-site, frameshift, indel, MNV, or unresolved
  bystander variants;
- whole-gene or multi-FASTA prompts;
- multiple questions, replicates, or treatment arms for the same gene;
- cross-assay score calibration or pooled correlation;
- consequence or amino-acid annotations in the prompt; and
- recomputing published functional scores from raw sequencing reads.
